src/main/flight/pid.c

```diff
@@ -436,28 +439,94 @@ STATIC_UNIT_TESTED FAST_CODE_NOINLINE float calcHorizonLevelStrength(void)
 // The impact is possibly slightly slower performance on F7/H7 but they have more than enough
 // processing power that it should be a non-issue.
 STATIC_UNIT_TESTED FAST_CODE_NOINLINE float pidLevel(int axis, const pidProfile_t *pidProfile, const rollAndPitchTrims_t *angleTrim,
-                                                        float currentPidSetpoint, float horizonLevelStrength)
+                                                        float rawSetpoint, float horizonLevelStrength, bool newRcFrame)
 {
     const float levelAngleLimit = pidProfile->levelAngleLimit;
     // calculate error angle and limit the angle to the max inclination
     // rcDeflection is in range [-1.0, 1.0]
-    float angle = levelAngleLimit * getLevelModeRcDeflection(axis);
+    float angleTarget = levelAngleLimit * getLevelModeRcDeflection(axis);
 #ifdef USE_GPS_RESCUE
-    angle += gpsRescueAngle[axis] / 100; // ANGLE IS IN CENTIDEGREES
+    angleTarget += gpsRescueAngle[axis] / 100; // ANGLE IS IN CENTIDEGREES
+#endif
+    const float currentAngle = (attitude.raw[axis] - angleTrim->raw[axis]) / 10.0f;
+    const float angleError = angleTarget - currentAngle;
+    const float angleErrorCorrection = angleError * pidRuntime.levelGain;
+    float currentAngleDerivative = pidRuntime.angleDerivative[axis];
+    if(newAttitudeData){
```

Comment:
If currentAttitude was filtered with the attitudeFilter, as suggested above, there would be no need for the `if(newAttitudeData)` check, and no need to filter the calculated D value since the input would already be smoothed.

Reply:
CurrentAngle data does not appear to update at PID loop rate. It updates at 1/2 the accelerometer sample rate in angle mode (~400Hz in my setup). Hence the need for if(newAttitudeData) otherwise the derivative term ends up as a heap of delta functions and is unusable.